market_maker_keeper/uniswapv2_market_maker_keeper.py

```diff
@@ -0,0 +1,441 @@
+            'amountAMin': amount_a_min,
+            'amountBMin': amount_b_min
+        }
+
+        remove_liquidity_eth_args = {
+            'liquidity': liquidity_to_remove,
+            'amountTokenMin': amount_a_min,
+            'amountETHMin': amount_b_min
+        }
+
+        if liquidity_to_remove > Wad(0):
+            self.logger.info(f"Removing {liquidity_to_remove} from Uniswap pool {self.uniswap.pair_address}")
+
+            if self.is_eth:
+                transact = self.uniswap.remove_liquidity_eth(remove_liquidity_eth_args, self.token_a).transact(
+                    gas_price=self.gas_price)
+            else:
+                transact = self.uniswap.remove_liquidity(remove_liquidity_args, self.token_a, self.token_b).transact(
+                    gas_price=self.gas_price)
+
+            if transact is not None and transact.successful:
+                gas_used = transact.gas_used
+                gas_price = Wad(self.web3.eth.getTransaction(transact.transaction_hash.hex())['gasPrice'])
+                tx_fee = Wad.from_number(gas_used) * gas_price
+
+                token_a_balance_after_remove = self.uniswap.get_account_token_balance(self.token_a)
+                token_a_removed = token_a_balance_after_remove - token_a_balance
+                if self.is_eth:
+                    eth_balance_after_remove = self.uniswap.get_account_eth_balance()
+                    eth_real_added = eth_balance_after_remove - eth_balance + tx_fee
+                    self.logger.info(f"Real Eth amount removed {eth_real_added} Real token amount "
+                                     f"removed {token_a_removed} at price {eth_real_added / token_a_removed}; "
+                                     f"tx fee used {tx_fee}"
+                                     f"with tx hash {transact.transaction_hash.hex()}")
+                else:
+                    token_b_balance_after_remove = self.uniswap.get_account_token_balance(self.token_b)
+                    token_b_removed = token_b_balance_after_remove - token_b_balance
+                    self.logger.info(f"Real {self.token_a.name} amount removed: {token_a_removed} "
+                                     f"Real {self.token_b.name} amount removed: {token_b_removed} "
+                                     f"tx fee used {tx_fee} "
+                                     f"with tx hash {transact.transaction_hash.hex()}")
+
+                return transact
+            else:
+                self.logger.warning(f"Failed to remove {liquidity_to_remove} liquidity of {self.uniswap.pair_address.address}")
+        else:
+            self.logger.info(f"No liquidity to remove")
+
+    def place_liquidity(self):
+        """
+        Add or remove liquidity depending upon the difference between Uniswap ratio and our external price feeds.
+
+        Pricing is expressed as a function of token_a balance / token_b balance
+        """
+
+        # feed_price = self.feed.get()[0]['price']
```

Comment:
shouldn't this read from price feed instead using hardcodded value?

Reply:
yes, would like to echo this. In order for Uniswap to work with our system the keeper will need to have some logic that handles our typical price feed input arg `--price-feed` not quite sure how `--uniswap-feed` works / will be different.

Most importantly, and you may already have this covered Mike (but didn't see anything in the below code), We need to be sure that if the price feed is giving Null values the bot stops placing orders and exits all liquidity for that pair

Reply:
Added price feed support. Also, I've added logic on line 414 in `determine_liquidity_action()` that will remove all liquidity on a null price feed.